Major update to alphabet module, add RNA

h-2 opened the pull request:
I took care to create a proper commit history so the changes should be evident and can be browsed individually. [don't destroy it by stashing :pray: ]

Mostly cleanup, style guide, minor fixes and clarification (private statics, noexcept, return ref after assignment...), but there is one "design decision": Rna alphabets are implemented in a way where they inherit from their dna counterparts and are assignable with them in both directions. Furthermore they are specified as having an identical internal representation. I think this makes, as it saves some code redundancy and also makes life easier for everyone.

@sarahet @eldariont I touched some of "your" files so you are very invited to review the changes!

Other API changes:
  * `from_char` and `from_integral` are now `assign_char` and `assign_integral`
     * reason: they are actually assign functions. Especially for the free function `from_`* is not intuitive.
  * `to_integral`, `assign_integral` and `underlying_integral_type` have been renamed to `to_rank`, `assign_rank` ... 
     * I was never too happy with integral and now I realised that chars are also integral so it doesn't make any sense to use this term here.

h-2 commented:
@marehr Since @rrahn is still out of office, can you review this? Then I will also review your other two PRs today!

marehr commented on include/seqan3/alphabet/composition.hpp:
```diff
@@ -175,41 +175,41 @@ struct alphabet_composition :
-            return to_integral(get<0>(*this)) +
-                   ((to_integral(get<idx + 1>(*this)) * cummulative_alph_sizes[idx]) + ...);
+            return to_rank(get<0>(*this)) +
+                   ((to_rank(get<idx + 1>(*this)) * cummulative_alph_sizes[idx]) + ...);
```
This remark has nothing to do with the PR

----
Why the if branch?

```c++
return ( (to_rank(get<idx + 1>(*this)) * cummulative_alph_sizes[idx]) + ... + to_rank(get<0>(*this)));
```

You could further simplify that (if cummulative_alph_sizes[0] = 1)
```c++
return ( (to_rank(get<idx>(*this)) * cummulative_alph_sizes[idx]) + ... + 0);

h-2 replied:
The thing is that the formula for computing this is
rank(0)
rank(1) * cum_alph_size(0)
rank(2) * cum_alph_size(1)
...
So we would need to make cum_alph_size size n + 1 which would make this function a little shorter, but would make the code more difficult to understand, because cum_alph_size[i] would not refer to the cummulative size of the i-th alphabet anymore. But it would also make the other function body simpler... I will think about it. thanks for the pointer!

marehr commented on include/seqan3/alphabet/concept.hpp:
```diff
@@ -47,42 +47,12 @@ namespace seqan3
-} // namespace seqan3::detail
-//! \publicsection
-
-// ------------------------------------------------------------------
```
Why is this not needed anymore?

h-2 replied:
Our real alphabet concept expects free functions, but for our own alphabet implementations we only provide members, because it is easier/cleaner to write, and people might want to use members, too. So for our alphabets to satisfy our own concept, we need to free function -> member wrappers.
Previously I defined a "internal_alphabet" to constrain these free functions. But it caused lots of confusion. Also it is unclear whether to put it into `detail`, than it is not clear to users how the free functions are constrained. Or to put it into `seqan3` than it adds more confusion for users what's an` internal alphabet concept`?
Another benefit of having individual constraints instead of an extra concept is, that some types can combine local/global overloads and it will work.

marehr commented on include/seqan3/alphabet/concept.hpp:
```diff
@@ -47,42 +47,12 @@ namespace seqan3
 //! Type metafunction that returns the `char_type` defined inside an alphabet type.
 template <typename alphabet_type>
-    requires detail::internal_alphabet_concept<alphabet_type>
+    requires requires (alphabet_type c) { typename alphabet_type::char_type; }
```
Why the code duplication and why did you remove the named concept?

h-2 replied:
There is no real duplication, every function now specifies it's own requirements.

marehr commented on include/seqan3/alphabet/nucleotide/dna4.hpp:
```diff
@@ -31,147 +31,159 @@
-        T,
-        U = T,
-        UNKNOWN = A
-    };
```
But, is know not `dna4{'a'}` possible again?

marehr replied:
nevermind

marehr commented on include/seqan3/alphabet/nucleotide/dna4.hpp:
```diff
@@ -260,6 +273,11 @@ struct dna4
+public:
+    //!\privatesection
+    //!\brief The data member.
+    internal_type _value;
```
you can do this? and it is still an aggregate type?

h-2 replied:
Yes, private and protected **static** members are allowed in pod types and aggregate types.

marehr commented on include/seqan3/alphabet/nucleotide/dna4.hpp:
```diff
@@ -269,9 +287,120 @@ constexpr dna4 dna4::T{internal_type::T};
+ * \relates dna4
+ *
+ * \attention
+ * Note that we recommend using @link dna4_vector @endlink instead of dna4_string in almost all situations.
```
`@link` and `@endlink` ?

h-2 replied:
I knew I forget some of the old syntax :sob:

marehr commented on include/seqan3/alphabet/nucleotide/dna4.hpp:
```diff
@@ -269,9 +287,120 @@ constexpr dna4 dna4::T{internal_type::T};
+ *~~~~~~~~~~~~~~~
+ *
+ * \attention
+ * All user-defined literals are in the namespace seqan3::literal!
```
Is this really user-defined from the perspective of your users?

h-2 replied:
I am currently unhappy about the `view` namespace anyway. Maybe we should discuss with @rrahn whether we get rid of it altogether? I will change the wording in any case.

marehr commented on include/seqan3/alphabet/nucleotide/dna4.hpp:
```diff
@@ -269,9 +287,120 @@ constexpr dna4 dna4::T{internal_type::T};
+namespace seqan3::literal
+{
+
+/*!\brief dna4 literal (returns @link dna4_vector @endlink)
```
`@link @endlink`

marehr commented on include/seqan3/alphabet/nucleotide/dna4.hpp:
```diff
@@ -269,9 +287,120 @@ constexpr dna4 dna4::T{internal_type::T};
+    return r;
+}
+
+/*!\brief dna4 string literal (returns @link dna4_string @endlink)
```
`@link`

marehr commented on include/seqan3/alphabet/nucleotide/dna4.hpp:
```diff
@@ -269,9 +287,120 @@ constexpr dna4 dna4::T{internal_type::T};
+ *     auto bax = "ACGTTA"_dna4s;
+ *~~~~~~~~~~~~~~~
+ *
+ * Please note the limitations of seqan3::dna4_string and consider using the @link operator""_dna4 @endlink instead.
```
`@link`

marehr commented on include/seqan3/alphabet/nucleotide/dna5.hpp:
```diff
@@ -271,9 +293,120 @@ constexpr dna5 dna5::N{internal_type::N};
+ *~~~~~~~~~~~~~~~
+ *
+ * \attention
+ * All user-defined literals are in the namespace seqan3::literal!
```
Same is this user-defined for our users?

marehr commented on include/seqan3/alphabet/nucleotide/dna5.hpp:
```diff
@@ -271,9 +293,120 @@ constexpr dna5 dna5::N{internal_type::N};
+namespace seqan3::literal
+{
+
+/*!\brief dna5 literal (returns @link dna5_vector @endlink)
```
`@link`

marehr commented on include/seqan3/alphabet/nucleotide/dna5.hpp:
```diff
@@ -271,9 +293,120 @@ constexpr dna5 dna5::N{internal_type::N};
+ * \relates dna5
+ *
+ * \attention
+ * Note that we recommend using @link dna5_vector @endlink instead of dna5_string in almost all situations.
```
`@link`

marehr commented on include/seqan3/alphabet/nucleotide/dna5.hpp:
```diff
@@ -271,9 +293,120 @@ constexpr dna5 dna5::N{internal_type::N};
+    return r;
+}
+
+/*!\brief dna5 string literal (returns @link dna5_string @endlink)
```
`@link`

marehr commented on include/seqan3/alphabet/nucleotide/nucl16.hpp:
```diff
@@ -120,79 +103,109 @@ struct nucl16
-    //! \privatesection
-    // conversion tables
+protected:
+    //!\privatesection
```
where does the `\publicsection` start again?

h-2 replied:
it is not necessary, because there is nothing following it than shall become public again (also the public keyword implicitly begins a public section again).

marehr commented on include/seqan3/alphabet/nucleotide/dna5.hpp:
```diff
@@ -30,147 +30,163 @@
-    //! \privatesection
-    // conversion tables
+protected:
+    //!\privatesection
```
where does the `\publicsection` start again?

marehr commented on include/seqan3/alphabet/nucleotide/dna4.hpp:
```diff
@@ -31,147 +31,159 @@
-    //! \privatesection
-    // conversion tables
+protected:
+    //!\privatesection
```
where does the `\publicsection` start again?

marehr commented on include/seqan3/alphabet/nucleotide/nucl16.hpp:
```diff
@@ -315,9 +333,140 @@ constexpr nucl16 nucl16::W{internal_type::W};
+    using dna16 = nucl16;
+    using rna16 = nucl16;
+    using dna = nucl16;
+    using rna = nucl16;
```
why is not `dna = dna4` and `rna = dna4`?

h-2 replied:
Ah, yes, this is also a design decision that I didn't document in the description of the PR. Since even is SeqAn2 people only used the bitcompression rarely we should use the most generic nucleotide type by default. (for regular containers sizeof(dna4 ) == sizeof(dna16) == 8bit)

marehr commented on include/seqan3/alphabet/nucleotide/nucl16.hpp:
```diff
@@ -315,9 +333,140 @@ constexpr nucl16 nucl16::W{internal_type::W};
+ * \relates nucl16
+ *
+ * \attention
+ * Note that we recommend using @link nucl16_vector @endlink instead of nucl16_string in almost all situations.
```
`@link`

marehr commented on include/seqan3/alphabet/nucleotide/nucl16.hpp:
```diff
@@ -315,9 +333,140 @@ constexpr nucl16 nucl16::W{internal_type::W};
+namespace seqan3::literal
+{
+
+/*!\brief nucl16 literal (returns @link nucl16_vector @endlink)
```
`@link`

marehr commented on include/seqan3/alphabet/nucleotide/nucl16.hpp:
```diff
@@ -315,9 +333,140 @@ constexpr nucl16 nucl16::W{internal_type::W};
+    using rna16 = nucl16;
+    using dna = nucl16;
+    using rna = nucl16;
+    //!\}
```
this is confusing, above is it within `/**/` and now in `//` can it be in the same type of braces.

h-2 replied:
`/*` is a multi-line comment, `//` is a single line comment. But for doxygen it has a different meaning also. In this case we need multi-line comments in the beginning, because the commands refer to the group and not the member. If you use single-line commands they instead refer to the individual using lines.
I know, it's not really great :neutral_face:

marehr commented on include/seqan3/alphabet/nucleotide/nucl16.hpp:
```diff
@@ -315,9 +333,140 @@ constexpr nucl16 nucl16::W{internal_type::W};
+ *~~~~~~~~~~~~~~~
+ *
+ * \attention
+ * All user-defined literals are in the namespace seqan3::literal!
```
user-defined?

marehr commented on include/seqan3/alphabet/nucleotide/nucl16.hpp:
```diff
@@ -315,9 +333,140 @@ constexpr nucl16 nucl16::W{internal_type::W};
+    return r;
+}
+
+/*!\brief nucl16 string literal (returns @link nucl16_string @endlink)
```
`@link`

marehr commented on include/seqan3/alphabet/nucleotide/nucl16.hpp:
```diff
@@ -315,9 +333,140 @@ constexpr nucl16 nucl16::W{internal_type::W};
+ *     auto bax = "ACGTTA"_nucl16s;
+ *~~~~~~~~~~~~~~~
+ *
+ * Please note the limitations of seqan3::nucl16_string and consider using the @link operator""_nucl16 @endlink instead.
```
This is a strange comment, because you use `operator""_nucl16``along with `seqan3::nucl16_string`

h-2 replied:
_nucl16 returns to nucl16_vector
_nucl16s returns nucl16_string
Since we don't recommend nucl16_string, we also recommend _nucl16 instead of _nucl16s. :bulb:

marehr commented on include/seqan3/alphabet/nucleotide/rna4.hpp:
```diff
@@ -0,0 +1,263 @@
+    /*!\name Read functions
+     * \{
+     */
+    //!\brief Return the letter as a character of char_type.
```
why not everything `//`

h-2 replied:
See above, the group stuff goes together in one comment block.

marehr commented on include/seqan3/alphabet/nucleotide/rna4.hpp:
```diff
@@ -0,0 +1,263 @@
+    }
+
+    //!\brief Assign from seqan3::dna4.
+    constexpr rna4 & operator=(dna4 const in) noexcept
```
why is this allowed but not the other way around?

marehr replied:
no const ref?

h-2 replied:
> why is this allowed but not the other way around?

The other way around already works automatically, because rna4 inherits dna4.

> no const ref?

dna4 is smaller than `&`

marehr commented on include/seqan3/alphabet/nucleotide/rna4.hpp:
```diff
@@ -0,0 +1,263 @@
+ *~~~~~~~~~~~~~~~
+ */
+
+struct rna4 : public dna4
```
why does it inherit from dna4

h-2 replied:
see description of this PR :smile:

marehr commented on include/seqan3/alphabet/nucleotide/rna4.hpp:
```diff
@@ -0,0 +1,263 @@
+ * \relates rna4
+ *
+ * \attention
+ * Note that we recommend using @link rna4_vector @endlink instead of rna4_string in almost all situations.
```
`@link`

marehr commented on include/seqan3/alphabet/nucleotide/rna4.hpp:
```diff
@@ -0,0 +1,263 @@
+namespace seqan3::literal
+{
+
+/*!\brief rna4 literal (returns @link rna4_vector @endlink)
```
`@link`

marehr commented on include/seqan3/alphabet/nucleotide/rna4.hpp:
```diff
@@ -0,0 +1,263 @@
+    return r;
+}
+
+/*!\brief rna4 string literal (returns @link rna4_string @endlink)
```
`@link`

marehr commented on include/seqan3/alphabet/nucleotide/rna5.hpp:
```diff
@@ -0,0 +1,266 @@
+
+    /*!\name Read functions
+     * \{
+     */
```
why not everything `//`

h-2 replied:
see above

marehr commented on include/seqan3/alphabet/nucleotide/rna4.hpp:
```diff
@@ -0,0 +1,263 @@
+constexpr rna4 rna4::C{internal_type::C};
+constexpr rna4 rna4::G{internal_type::G};
+constexpr rna4 rna4::U{internal_type::U};
+constexpr rna4 rna4::T{rna4::U};
```
why is there a T?

h-2 replied:
Rna can be assigned a T and Dna can be assigned a U. They are functionally the same. Only printing to char yields a different result.

marehr commented on include/seqan3/alphabet/nucleotide/rna5.hpp:
```diff
@@ -0,0 +1,266 @@
+    static const rna5 A;
+    static const rna5 C;
+    static const rna5 G;
+    static const rna5 T;
```
why is there a T?

marehr commented on include/seqan3/alphabet/nucleotide/rna5.hpp:
```diff
@@ -0,0 +1,266 @@
+ * \relates rna5
+ *
+ * \attention
+ * Note that we recommend using @link rna5_vector @endlink instead of rna5_string in almost all situations.
```
`@link`

marehr commented on include/seqan3/alphabet/nucleotide/rna5.hpp:
```diff
@@ -0,0 +1,266 @@
+namespace seqan3::literal
+{
+
+/*!\brief rna5 literal (returns @link rna5_vector @endlink)
```
`@link`

marehr commented on include/seqan3/alphabet/nucleotide/rna5.hpp:
```diff
@@ -0,0 +1,266 @@
+    return r;
+}
+
+/*!\brief rna5 string literal (returns @link rna5_string @endlink)
```
`@link`

marehr commented on include/seqan3/alphabet/nucleotide/rna5.hpp:
```diff
@@ -0,0 +1,266 @@
+ *     auto bax = "ACGTTA"_rna5s;
+ *~~~~~~~~~~~~~~~
+ *
+ * Please note the limitations of seqan3::rna5_string and consider using the @link operator""_rna5 @endlink instead.
```
`@link`

marehr commented on include/seqan3/alphabet/quality/aliases.hpp:
```diff
@@ -50,25 +50,31 @@
-static_assert(seqan3::quality_concept<seqan3::dna4q>);
-static_assert(seqan3::detail::internal_quality_concept<seqan3::dna4q>);
+static_assert(seqan3::nucleotide_concept<seqan3::dna4q>);
 static_assert(sizeof(seqan3::dna4q) == 2);
```
what is `static_assert(sizeof(seqan3::nucl16q) == ??);`

h-2 replied:
It's a guarantee that the implementation doesn't have any size overhead over the individual alphabets.

marehr commented on include/seqan3/alphabet/quality/composition.hpp:
```diff
@@ -166,6 +167,16 @@ quality_composition(sequence_alphabet_type &&, quality_alphabet_type &&)
+//!\brief Since seqan3::quality_composition wraps a nucleotide alphabet it is also one.
+template <typename sequence_alphabet_type, typename quality_alphabet_type>
+struct is_nucleotide<quality_composition<sequence_alphabet_type, quality_alphabet_type>> : public std::true_type
+{};
```
Does it also satisfy the `quality_concept`?

h-2 replied:
See the definition of  `quality_concept`: it requires the `to_phred` `assign_phred` ... stuff. Since this is provided, the type satisfies quality_concept (this is tested in another file).
`nucleotide_concept` is special, in that it needs to have an explicit metafunction set to true. That's because structurally it doesn't require more of the type, than e.g. aminoacid, i.e. no different member functions or types, but it represents something different (and it makes sense to restrict certain functions to one or the other).

marehr commented on include/seqan3/alphabet/quality/composition.hpp:
```diff
@@ -166,6 +167,16 @@ quality_composition(sequence_alphabet_type &&, quality_alphabet_type &&)
+
 #ifndef NDEBUG
 // contains alphabet tests for this class:
 #include <seqan3/alphabet/quality/aliases.hpp>
```
I would not this like that, I would the test cases here directly. It more easy to understand what you want to assert.

```c++
#include <seqan3/alphabet/nucleotide/dna4.hpp>
#include <seqan3/alphabet/quality/illumina18.hpp>

static_assert(seqan3::nucleotide_concept<seqan3::quality_composition<seqan3::dna4, seqan3::illumina18>>);
static_assert(seqan3::quality_concept<seqan3::quality_composition<seqan3::dna4, seqan3::illumina18>>);
```

h-2 replied:
ok

marehr commented on include/seqan3/alphabet/quality/concept.hpp:
```diff
@@ -85,12 +85,12 @@ template <typename alphabet_type>
 template <typename alphabet_type>
     requires detail::internal_quality_concept<alphabet_type>
-    constexpr alphabet_type from_phred(alphabet_type & c, char const in)
+    constexpr alphabet_type assign_phred(alphabet_type & c, char const in)
```
Do we need those meta functions? `&` in return type?

h-2 replied:
they are functions, not meta functions... Of course we need them :)
Yes, it should be `&` on returning, but the entire quality submodule is WIP by @mariehoffmann (see #39) and I don't want to edit everything there now. If we get the rest of the alphabet module into a good state, I can then explain the design decisions to @mariehoffmann so that the quality submodule an be adapted in a similar manner (e.g. remove the internal_alphabet ..).

marehr commented on include/seqan3/alphabet/quality/concept.hpp:
```diff
@@ -109,8 +109,8 @@ concept bool quality_concept = requires(q quality)
-    //! requires additionally public getter and setter for integral phred score representation
-    { from_phred(quality, typename q::integral_type{}) } -> q;
+    //! requires additionally public getter and setter for rank phred score representation
+    { assign_phred(quality, typename q::rank_type{}) } -> q;
```
Why the Metafunctions? If `rank_type` is depend on `q`, why not directly use `q.assign_phred()`?

h-2 replied:
concepts that are used in interfaces and that might be satisfied by external types should only require free functions, not member functions. At least everywhere where we get to define the concept ourselves.

marehr commented on test/alphabet/nucleotide/dna4_test.cpp:
```diff
@@ -0,0 +1,327 @@
+    EXPECT_EQ(dna4::UNKNOWN.to_char(), 'A');
+}
+
+TEST(dna4, to_char_free)
```
to_char_free?

marehr replied:
I think this testcase should belong in a different scope

h-2 replied:
Hm, it's part of the alphabet concept and thereby part of the "official interface". I don't think it hurts to test it here.

marehr commented on test/alphabet/nucleotide/dna4_test.cpp:
```diff
@@ -0,0 +1,327 @@
+    EXPECT_EQ(to_char(dna4::UNKNOWN), 'A');
+}
+
+TEST(dna4, to_rank_member)
```
this test a wrapper, and should be tested in a different test case

marehr commented on test/alphabet/nucleotide/dna4_test.cpp:
```diff
@@ -0,0 +1,327 @@
+    EXPECT_EQ(dna4::UNKNOWN.to_rank(), 0);
+}
+
+TEST(dna4, to_rank_free)
```
this test a wrapper, and should be tested in a different test case

marehr commented on test/alphabet/nucleotide/dna4_test.cpp:
```diff
@@ -0,0 +1,327 @@
+    EXPECT_EQ(t0.assign_char('C'), dna4::C);
+}
+
+TEST(dna4, assign_char_free)
```
this test a wrapper, and should be tested in a different test case

marehr commented on test/alphabet/nucleotide/dna4_test.cpp:
```diff
@@ -0,0 +1,327 @@
+    EXPECT_EQ(t0.assign_rank(1), dna4::C);
+}
+
+TEST(dna4, assign_rank_free)
```
this test a wrapper, and should be tested in a different test case

marehr commented on test/alphabet/nucleotide/dna5_test.cpp:
```diff
@@ -31,90 +31,314 @@
 }
 
-TEST(dna5_test, test_implicit_inner_type_compatibility)
+TEST(dna5, to_char_free)
```
this test a wrapper, and should be tested in a different test case

marehr commented on test/alphabet/nucleotide/dna5_test.cpp:
```diff
@@ -31,90 +31,314 @@
 }
 
-TEST(dna5_test, test_static_cast)
+TEST(dna5, to_rank_free)
```
this test a wrapper, and should be tested in a different test case

marehr commented on test/alphabet/nucleotide/dna5_test.cpp:
```diff
@@ -31,90 +31,314 @@
 }
 
-TEST(dna5_test, test_to_char)
+TEST(dna5, assign_char_free)
```
this test a wrapper, and should be tested in a different test case

marehr commented on test/alphabet/nucleotide/dna5_test.cpp:
```diff
@@ -31,90 +31,314 @@
-    EXPECT_EQ(dna5{}.from_char('g'), dna5{dna5::G});
-    EXPECT_EQ(dna5{}.from_char('t'), dna5{dna5::T});
-    EXPECT_EQ(dna5{}.from_char('n'), dna5{dna5::N});
+TEST(dna5, assign_rank_free)
```
this test a wrapper, and should be tested in a different test case

marehr commented on test/alphabet/nucleotide/nucl16_test.cpp:
```diff
@@ -30,141 +30,233 @@
-// ============================================================================
-
-#include <array>
+// ==========================================================================
```
the length does not match `// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH` anymore :grin:

marehr commented on test/alphabet/nucleotide/nucl16_test.cpp:
```diff
@@ -30,141 +30,233 @@
 
-TEST(alphabet_nucleotides_nucl16_test, copy_ctr)
+// copy construction
+TEST(nucl16, cp_ctr)
```
why this so short names

h-2 replied:
save bytes, save energy, save the planet :seedling: 
No seriously, since it even has a comment before, I think it is very clear, or no?

marehr commented on test/alphabet/nucleotide/nucl16_test.cpp:
```diff
@@ -30,141 +30,233 @@
 {
-    constexpr nucl16 nucleotide1{nucl16::B};
-    constexpr nucl16 nucleotide2{nucleotide1};
+    nucl16 t1{nucl16::C};
```
Why so short variable names?

I think someone should be able to directly copy from test cases with somehow meaningful names.

`t0` you can't guess what should be.

h-2 replied:
I think the tests are super short and rather intuitive. Whether you call a variable `a` or `i` or `t0` or `t1` doesn't make much of difference.

marehr commented on test/alphabet/nucleotide/nucl16_test.cpp:
```diff
@@ -30,141 +30,233 @@
+TEST(nucl16, assign_char_member)
+{
+    nucl16 t0;
+    for (size_t i = 0; i < 16; ++i)
```
I think you should at least show once how what you are generalizing.

```c++
t0.assign_char('a');
EXPECT_EQ(t0, nucl16::A);
```

marehr replied:
I'm exhausted. Hopefully someone else can take the rest, or I will do the rest next week.

h-2 replied:
ok

h-2 replied:
Thanks!

h-2 commented:
@rrahn @marehr since this PR is getting bigger and bigger and both of you are on sick-leave, I am merging this now. I have addressed most of the points from @marehr  and this is blocking a lot of other planned changes...
